File: app/src/main/java/com/aspirasoft/huntrek/activity/ActivityCollect.java
package com.aspirasoft.huntrek.activity;

import android.app.Activity;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.Switch;
import android.widget.TextView;
import com.aspirasoft.huntrek.R;
import com.aspirasoft.huntrek.bo.UserManager;
import com.aspirasoft.huntrek.model.User;
import com.aspirasoft.huntrek.view.LoadingAnimation;
import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.interfaces.DraweeController;
import com.facebook.drawee.view.SimpleDraweeView;
import com.unity3d.player.UnityPlayer;

public class ActivityCollect extends Activity implements View.OnClickListener {

    protected UnityPlayer mUnityPlayer;
    private FrameLayout mUnitPlayerContainer;
    private LoadingAnimation mLoadingView;

    private ImageButton mSettingsButton;
    private Switch mARSwitch;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        View decorView = getWindow().getDecorView();
        int uiOptions = View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN;
        decorView.setSystemUiVisibility(uiOptions);
        setContentView(R.layout.activity_collect);

        // Check that user is logged in
        User currentUser = UserManager.getInstance().getActiveUser();
        if (currentUser == null) {
            finish();
            return;
        }

        // Assign views
        mUnitPlayerContainer = findViewById(R.id.unity_player_view);
        mSettingsButton = findViewById(R.id.button_menu);
        mLoadingView = findViewById(R.id.pauseScreen);
        mARSwitch = findViewById(R.id.ARSwitch);

        // Assign click listeners
        mSettingsButton.setOnClickListener(this);

        // Configure unity player
        mUnityPlayer = new UnityPlayer(this);
        mUnitPlayerContainer.addView(
                mUnityPlayer, new FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT
                ));

        // Display 'loading' screen
        mLoadingView.setVisibility(View.VISIBLE);
        mUnityPlayer.requestFocus();

        // Schedule 'loading' screen to auto-hide after 3.5 seconds
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        mLoadingView.setVisibility(View.GONE);
                    }
                });
            }
        }, 3500L);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        // To support deep linking, we need to make sure that the client can get access to
        // the last sent intent. The clients access this through a JNI api that allows them
        // to get the intent set on launch. To update that after launch we have to manually
        // replace the intent with the one caught here.
        setIntent(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mUnityPlayer.start();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mUnityPlayer.pause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mUnityPlayer.resume();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mUnityPlayer.stop();
    }

    @Override
    protected void onDestroy() {
        mUnityPlayer.quit();
        super.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mUnityPlayer.lowMemory();
    }

    @Override
    public void onTrimMemory(int level) {
        super.onTrimMemory(level);
        if (level == TRIM_MEMORY_RUNNING_CRITICAL) {
            mUnityPlayer.lowMemory();
        }
    }

    // This ensures the layout will be correct.
    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mUnityPlayer.configurationChanged(newConfig);
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        mUnityPlayer.windowFocusChanged(hasFocus);
    }

    // For some reason the multiple keyevent type is not supported by the ndk.
    // Force event injection by overriding dispatchKeyEvent().
    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_MULTIPLE)
            return mUnityPlayer.injectEvent(event);
        return super.dispatchKeyEvent(event);
    }

    // Pass any events not handled by (unfocused) views straight to UnityPlayer
    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        return mUnityPlayer.injectEvent(event);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        return mUnityPlayer.injectEvent(event);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return mUnityPlayer.injectEvent(event);
    }

    /*API12*/
    public boolean onGenericMotionEvent(MotionEvent event) {
        return mUnityPlayer.injectEvent(event);
    }

    @Override
    public void onClick(View v) {
        if (v.equals(mSettingsButton)) {
            v.setVisibility(View.GONE);
            UnityPlayer.UnitySendMessage("Treasure Chest", "Open", "open");

            DraweeController controller =
                    Fresco.newDraweeControllerBuilder()
                            .setUri("res:///" + R.drawable.fireworks)
                            .setAutoPlayAnimations(true)
                            .build();

            final SimpleDraweeView mFallingCoinsView = findViewById(R.id.falling_coins);
            mFallingCoinsView.setController(controller);

            Handler mHandler = new Handler();

            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mFallingCoinsView.setVisibility(View.VISIBLE);
                        }
                    });
                }
            }, 1500L);

            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Intent i = getIntent();
                            int coins = i.getIntExtra("ChestValue", 0);

                            TextView coinsView = findViewById(R.id.coins_collected);
                            coinsView.setText("Coins: " + coins);
                            coinsView.setVisibility(View.VISIBLE);
                        }
                    });
                }
            }, 2000L);

            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            finish();
                            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
                        }
                    });
                }
            }, 4000L);
        }
    }

    public void escape(View view) {
        finish();
        overridePendingTransition(0, 0);
    }

}

File: app/src/main/java/dev/aspirasoft/huntit/ui/activity/ActivityCollect.kt
package dev.aspirasoft.huntit.ui.activity

import android.content.ComponentCallbacks2
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView
import com.facebook.drawee.backends.pipeline.Fresco
import com.facebook.drawee.interfaces.DraweeController
import com.facebook.drawee.view.SimpleDraweeView
import com.unity3d.player.UnityPlayer
import dev.aspirasoft.huntit.R
import dev.aspirasoft.huntit.data.repo.AuthRepository
import kotlinx.android.synthetic.main.activity_collect.*

class ActivityCollect : FullScreenActivity(), View.OnClickListener {

    private lateinit var mUnityPlayer: UnityPlayer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_collect)

        // Check that user is logged in
        if (!AuthRepository.isSignedIn) {
            finish()
            return
        }

        // Assign click listeners
        settingsButton.setOnClickListener(this)

        // Configure unity player
        mUnityPlayer = UnityPlayer(this)
        val unityPlayerView = findViewById<FrameLayout>(R.id.unityPlayerView)
        unityPlayerView.addView(mUnityPlayer, FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        ))
        mUnityPlayer.requestFocus()

        // Display 'loading' screen
        pauseScreen.visibility = View.VISIBLE

        // Schedule 'loading' screen to auto-hide after 3.5 seconds
        Handler().postDelayed({ runOnUiThread { pauseScreen.visibility = View.GONE } }, 3500L)
    }

    /**
     * To support deep linking, we need to make sure that the client can get access to
     * the last sent intent. The clients access this through a JNI api that allows them
     * to get the intent set on launch. To update that after launch we have to manually
     * replace the intent with the one caught here.
     */
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
    }

    override fun onStart() {
        super.onStart()
        mUnityPlayer.start()
    }

    override fun onPause() {
        super.onPause()
        mUnityPlayer.pause()
    }

    override fun onResume() {
        super.onResume()
        mUnityPlayer.resume()
    }

    override fun onStop() {
        super.onStop()
        mUnityPlayer.stop()
    }

    override fun onDestroy() {
        mUnityPlayer.quit()
        super.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mUnityPlayer.lowMemory()
    }

    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
        if (level == ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL) {
            mUnityPlayer.lowMemory()
        }
    }

    /**
     *  This ensures the layout will be correct.
     */
    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        mUnityPlayer.configurationChanged(newConfig)
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        mUnityPlayer.windowFocusChanged(hasFocus)
    }

    /**
     * For some reason the multiple keyevent type is not supported by the ndk.
     * Force event injection by overriding dispatchKeyEvent().
     */
    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        return if (event.action == KeyEvent.ACTION_MULTIPLE) mUnityPlayer.injectEvent(event) else super.dispatchKeyEvent(
            event)
    }

    /**
     * Pass any events not handled by (unfocused) views straight to UnityPlayer
     */
    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        return mUnityPlayer.injectEvent(event)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return mUnityPlayer.injectEvent(event)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return mUnityPlayer.injectEvent(event)
    }

    /**
     * For API12
     */
    override fun onGenericMotionEvent(event: MotionEvent): Boolean {
        return mUnityPlayer.injectEvent(event)
    }

    override fun onClick(v: View) {
        if (v == settingsButton) {
            v.visibility = View.GONE
            UnityPlayer.UnitySendMessage("Treasure Chest", "Open", "open")
            val controller: DraweeController = Fresco.newDraweeControllerBuilder()
                .setUri("res:///" + R.drawable.fireworks)
                .setAutoPlayAnimations(true)
                .build()
            val mFallingCoinsView: SimpleDraweeView = findViewById(R.id.falling_coins)
            mFallingCoinsView.controller = controller
            val mHandler = Handler()
            mHandler.postDelayed({ runOnUiThread { mFallingCoinsView.visibility = View.VISIBLE } }, 1500L)
            mHandler.postDelayed({
                runOnUiThread {
                    val i = intent
                    val coins = i.getIntExtra("ChestValue", 0)
                    val coinsView = findViewById<TextView>(R.id.coins_collected)
                    coinsView.text = "Coins: $coins"
                    coinsView.visibility = View.VISIBLE
                }
            }, 2000L)
            mHandler.postDelayed({
                runOnUiThread {
                    finish()
                    overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
                }
            }, 4000L)
        }
    }

    fun escape(view: View?) {
        finish()
        overridePendingTransition(0, 0)
    }

}